README.md
# How Django Signals Work - Stand-alone Project 

```

In Django, signals allow certain senders 
to inform a set of receivers 
that specific actions have occurred. 

Django signals are used to send and 
receive specific essential information 
whenever a data model is 
saved, changed, or even removed.

```
## What exactly is this project?
```
It's a standalone Django Project
Running a simple How Signals Works 
in Django
```

Based: Dennis Ivy [vid](https://youtu.be/Kc1Q_ayAeQk)


This is powered by [Django](https://www.djangoproject.com/) , a web framework for perfectionists with deadlines!

Django is a full-stack, open-source Python framework designed for efficient web development.

Why Django?
```
Django is a high-level Python web framework 
that encourages rapid development and clean, pragmatic design. 

Built by experienced developers, it takes care of much 
of the hassle of web development, so you can focus on writing 
your app without needing to reinvent the wheel. 
It’s free and open source.
```
You're more then welcome to visit my Web Pages:

[Jungletronics](https://medium.com/jungletronics) (Arduino, RPi, PIC, Eagle, Blender, Unity3D, Pixy, and more)
[KidsTronics](https://medium.com/kidstronics) (MIT App Inventor, LEGO, Arduino For Kids, and more)

And my [Youtube Channel Playlist](https://www.youtube.com/playlist?list=PLK3PeNcUzb8TwZuXZJgREj5nDbQxRLW_a)

Or my old Atlassian,Inc. Repo: https://bitbucket.org/gilj3/

###	Quick Start
To get this project up and running locally on your computer:

Set up the Python development environment. We recommend using a Python virtual environment.
Assuming you have Python setup, run the following commands (if you're on Windows you may use py or python-3 instead of python to start Python):

### Check list:  
[Would you like some help?](https://youtu.be/tr_2k87AdYQ)
Read [hystory.txt](sygnal_demo/history.txt) file

- [ ] pip3 install -r requirements.txt
- [ ] python3 manage.py makemigrations
- [ ] python3 manage.py makemigrations
- [ ] python3 manage.py migrate
- [ ] python3 manage.py collectstatic
- [ ] python3 manage.py test # Run the standard tests. These should all pass.
- [ ] python3 manage.py createsuperuser # Create a superuser
- [x] python3 -m venv djangoEnv  
- [x] djangoEnv\Scripts\activate 
- [x] python3 -m pip install --upgrade pip
- [x] python3 -m pip install django 
- [x] python3 -m django --version  
- [x] python3 -m pip install django-crispy-forms 
- [x] python3 -m pip install Pillow
- [x] python3 manage.py runserver
- [x] Open a browser to http://127.0.0.1:8000/admin/ to open the admin site
- [x] Create a superuser see console output ;b

### License

[![License: CC BY-NC-ND 3.0](https://img.shields.io/badge/License-CC%20BY--NC--ND%203.0-lightgrey.svg)](https://creativecommons.org/licenses/by-nc-nd/3.0/)


sygnal_demo/history.txt
mkdir django_signals
cd django_signals
python -m venv ENV
source ENV/bin/activate
(source /home/j3/django_signals/ENV/bin/activate)
pip install django
python manage.py createproject signals
cd signal
python manage.py createapp base
cd base/model.py
Type these lines & hit save:
----------------------------------------
from django.db import models                   
from django.contrib.auth.models import User    
from django.db.models.signals import post_save 


class Profile(models.Model):    
               
     user = models.OneToOneField(User, on_delete=models.CASCADE, blank=True, null=True) 
     first_name = models.CharField(max_length=200, blank=True, null=True) 
     last_name = models.CharField(max_length=200, blank=True, null=True) 
     phone = models.CharField(max_length=200, blank=True, null=True) 
     
     def __str__(self):                                            
         return str(self.user) 
'''

    def create_profile(sender, instance, created, **kwargs):
        if created:
            Profile.objects.create(user=instance)
            print('****************')
            print('Profile created!')
            print('****************')

    post_save.connect(create_profile, sender=User)

    def update_profile(sender, instance, created, **kwargs):
        if created is False:
            instance.profile.save()
            print('****************')
            print('Profile Updated!')
            print('****************')

    post_save.connect(update_profile, sender=User)
    
'''
----------------------------------------
This model connect Django User + my Profile ;)  
Now Go to admin.py:
----------------------------------------
from django.contrib import admin
from base.models import Profile

admin.site.register(Profile)
----------------------------------------
Go to apps.py
----------------------------------------
from django.apps import AppConfig


class BaseConfig(AppConfig):
    # default_auto_field = 'django.db.models.BigAutoField'
    name = 'base'

    def ready(self):
        import base.signals

----------------------------------------
Create & Go to signals.py:
----------------------------------------
from django.contrib.auth.models import User
from base.models import Profile
from django.db.models.signals import post_save
from django.dispatch import receiver


@receiver(post_save, sender=User)
def create_profile(sender, instance, created, **kwargs):
    if created:
        Profile.objects.create(user=instance)
        print('****************')
        print('Profile created!')
        print('****************')


# post_save.connect(create_profile, sender=User)

@receiver(post_save, sender=User)
def update_profile(sender, instance, created, **kwargs):
    if created is False:
        instance.profile.save()
        print('****************')
        print('Profile Updated!')
        print('****************')

----------------------------------------
python manage.py makemigrations
python manage.py migrate
pip freeze > requirement.txt
code .
exit
Inside vscode:
Go to /admin
Create & Login as superuser
create/ update a user
SEE Profile...

Output:
--------------------------
[05/Apr/2023 02:25:04] "GET /admin/auth/user/add/ HTTP/1.1" 200 7383
[05/Apr/2023 02:25:04] "GET /admin/jsi18n/ HTTP/1.1" 200 3195
****************
Profile created!
****************
[05/Apr/2023 02:25:35] "POST /admin/auth/user/add/ HTTP/1.1" 302 0
[05/Apr/2023 02:25:35] "GET /admin/auth/user/23/change/ HTTP/1.1" 200 16008
[05/Apr/2023 02:25:35] "GET /admin/jsi18n/ HTTP/1.1" 200 3195
****************
Profile Updated!
****************
[05/Apr/2023 02:26:01] "POST /admin/auth/user/23/change/ HTTP/1.1" 302 0
[05/Apr/2023 02:26:01] "GET /admin/auth/user/ HTTP/1.1" 200 8968
[05/Apr/2023 02:26:01] "GET /admin/jsi18n/ HTTP/1.1" 200 3195
--------------------------
There you have!
Hope this will be helpful!
j3
March, 2023


sygnal_demo/base/migrations/0001_initial.py
# Generated by Django 3.2.18 on 2023-04-04 20:32

from django.conf import settings
from django.db import migrations, models
import django.db.models.deletion


class Migration(migrations.Migration):

    initial = True

    dependencies = [
        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
    ]

    operations = [
        migrations.CreateModel(
            name='Profile',
            fields=[
                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('first_name', models.CharField(blank=True, max_length=200, null=True)),
                ('last_name', models.CharField(blank=True, max_length=200, null=True)),
                ('phone', models.CharField(blank=True, max_length=200, null=True)),
                ('user', models.OneToOneField(blank=True, null=True, on_delete=django.db.models.deletion.CASCADE, to=settings.AUTH_USER_MODEL)),
            ],
        ),
    ]
